### PR TITLE
Csiro/mk uri

### PR DESCRIPTION
**Pull request by petewa**
### Moved `mk_uri` function from `ingest.py` to Writer Driver

Before `mk_uri` was set by the ingest specifically for the NetCDF driver
Now `mk_uri` can be set by the Writer Driver in a form that it understands.

This also avoids the need for multiple writer endpoints in `setup.py`, should the driver supports multiple protocols/schemes.

### Proposed changes

- Move `mk_uri` from ingest.py into Writer Driver
- Change `write_dataset_to_storage` parameter from `filename` to `file_uri`

 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: docs/architecture/driver.rst [hunk 1]
```diff
@@ -113,7 +113,14 @@ Example code to implement a writer driver
         def format(self):
             return ''  # Format that this writer supports
 
-        def write_dataset_to_storage(self, dataset, filename,
+        def mk_uri(self, file_path, storage_config):
+            """
+            Constructs a uri from the file_path and storage config.
```
Comment (Kirill888) [inline docs/architecture/driver.rst]: Can we expand a bit on what `file_path` and `storage_config` parameters are?
Where they are coming from, what type they are.
What should a "typical" implementation that writes to local filesystem should return from this function.

File: datacube/drivers/netcdf/driver.py [hunk 2]
```diff
@@ -39,15 +42,26 @@ def format(self):
     def uri_scheme(self):
         return PROTOCOL
 
-    def write_dataset_to_storage(self, dataset, filename,
+    def mk_uri(self, file_path, storage_config):
+        """
+        Constructs a uri from the file_path and storage config.
+        resource.
+        """
+        driver_alias = storage_config['driver']
```
Comment (Kirill888) [inline datacube/drivers/netcdf/driver.py]: For this driver, I think it should always return `normalise_path(file_path).as_uri()` and simply ignore `storage_config` part.

File: docs/about/whats_new.rst [hunk 3]
```diff
@@ -9,6 +9,7 @@ v1.8.1 (???)
 ============
 
 - Added ``updated`` column for trigger based tracking of database row updates in PostgreSQL. (:pull:`951`)
+- Moved ``mk_uri`` from ``ingest.py`` into Writer Driver. (:pull:`960`)
```
Comment (Kirill888) [inline docs/about/whats_new.rst]: Probably worthwhile making this more verbose, something like "Changes to writer driver API. Driver is now responsible for constructing output URIs from user configuration."

**Review by Kirill888 — APPROVED** [review]
Thanks :+1:

**Comment by petewa** [comment]
Thank you Kirill for going through the PR.

**Review by Kirill888 — APPROVED** [review]

**Pull request merged**